### PR TITLE
MRG: Convergence

### PR DESCRIPTION
**Pull request by pavanramkumar**
A few things to note (see review comments):

- The criterion based on (relative) parameter change per iteration is sufficient for all tests to pass and speed up all examples despite strict default tolerance (`tol=1e-6`).

- However, I am leaving the gradient norm based criterion in the code for readability sake. My main concern is that to make the gradient-based convergence work well, we have to calculate the projected gradient (i.e. what remains after the proximal operator has been applied) and then take the norm on it. Unfortunately that required more digging into the implementation of `_prox` for all corner cases and I ran out of time. So I would recommend removing this criterion entirely for this PR.

- As a consequence of this PR, one test fails: `test_random_state_consistency`. However the failure is easily understood and can be traced back to cold (random) vs. warm initialization of beta. I've recommended a modification to the test: use `assert_allclose` instead of `assert_equal` (commented out for now; I wanted to demo the failed test first)

### CODE REVIEW

File: pyglmnet/pyglmnet.py [hunk 1]
```diff
@@ -812,33 +812,34 @@ def fit(self, X, y):
...
             if self.solver == 'batch-gradient':
                 grad = _grad_L2loss(self.distr,
                                     alpha, self.Tau,
                                     reg_lambda, X, y, self.eta,
                                     beta, self.fit_intercept)
-                # Converged if the norm(gradient) < tol
-                if (t > 1) and (np.linalg.norm(grad) < tol):
-                    msg = ('\tConverged in {0:d} iterations'.format(t))
+
+                # Convergence by gradient norm tolerance
```
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: I would recommend removing this criterion entirely

File: pyglmnet/pyglmnet.py [hunk 2]
```diff
@@ -851,6 +852,14 @@ def fit(self, X, y):
                 if self.fit_intercept:
                     ActiveSet[0] = 1.
 
+            # Convergence by relative parameter change tolerance
```
Comment (pavanramkumar) [inline pyglmnet/pyglmnet.py]: This criterion on its own is sufficient for all tests to pass

File: tests/test_pyglmnet.py [hunk 3]
```diff
@@ -421,6 +420,7 @@ def test_random_state_consistency():
     assert_array_equal(ypred_a, ypred_b)
     # Consistency between different run of the same model
     assert_array_equal(ypred_b, ypred_c)
+    # assert_allclose(ypred_b, ypred_c, atol=1e-5, rtol=1e-4)
```
Comment (pavanramkumar) [inline tests/test_pyglmnet.py]: I recommend this modification instead of the previous line

**Comment by pavanramkumar** [comment]
closes #319

File: pyglmnet/pyglmnet.py (hunk 2 above)
Comment (jasmainak) [reply pyglmnet/pyglmnet.py]: Do the convergence warnings go away once we switch to this criteria?

**Comment by jasmainak** [comment]
@pavanramkumar our current tests don't test if we stop our iterations too late. Perhaps we should have a test where we know that it converges in less than x iterations (e.g., with scikit-learn data)?

```py
assert glm.n_iter_ < 15
```

or something like that?

File: pyglmnet/pyglmnet.py (hunk 2 above)
Comment (pavanramkumar) [reply pyglmnet/pyglmnet.py]: yes

**Comment by pavanramkumar** [comment]
> @pavanramkumar our current tests don't test if we stop our iterations too late. Perhaps we should have a test where we know that it converges in less than x iterations (e.g., with scikit-learn data)?
> 
> ```python
> assert glm.n_iter_ < 15
> ```
> 
> or something like that?

The problem with this is that we cannot easily distinguish good and bad failures. Some optimization loops with warm restarts legit converge in 10 iterations. So I'd say let's wait to improve this

**Comment by pavanramkumar** [comment]
@jasmainak tests pass now. good to merge?

File: pyglmnet/pyglmnet.py [hunk 4]
```diff
@@ -812,33 +812,25 @@ def fit(self, X, y):
         # Iterative updates
         for t in range(0, self.max_iter):
             self.n_iter_ += 1
-
+            beta_old = deepcopy(beta)
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: why is this necessary?

Comment (jasmainak) [reply pyglmnet/pyglmnet.py]: actually you should do `beta.copy()`

**Comment by jasmainak** [comment]
@pavanramkumar I'm merging this so I can use it for some testing in my PR. I'll update the what's new in my PR
